ESQL: Pre-allocate rows in TopNOperator

alex-spies opened the pull request:
Starting with empty rows and growing them causes lots of allocations and thus bad performance in case of many large fields being contained in the rows.

Instead, use the previously encountered row to estimate the size of the next row.

This reduces the response time of the request sent in HeapAttackIT.testFetchManyBigFields from around 450ms to about 350 ms (22% improvement).

alex-spies committed:
Pre-allocate rows in TopNOperator

Starting with empty rows and growing them causes lots of allocations and
thus bad performance in case of many large fields being contained in the
rows.

Instead, use the previously encountered row to estimate the size of the
next row.

This reduces the response time of the request sent in
HeapAttackIT.testFetchManyBigFields from around 450ms to about 350 ms
(22% improvement).

alex-spies commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/topn/TopNOperator.java:
```diff
@@ -273,6 +276,12 @@ public String describe() {
+     * Esp. in case of many fields, a row's values can become quite large; growing from an empty row every time is expensive due
+     * to lots of allocations. Instead, keep track of the last row's actual size and use that to pre-allocate memory for the next row.
+     */
+    private int lastSeenRowValuesSize = 32;
```
We could do the same for the row's search keys, but I'm not sure it's worth it.

For this to matter, there'd have to be a) many search keys with b) large values (e.g. sorting by 100 keyword fields). In this case, the bottleneck may be somewhere else entirely.

What do you think @nik9000 ?

nik9000 replied:
I'd do it. I guess I figure "doing it twice" doesn't really add an cognitive overhead.

One thing that's neat - we could report this in the "status" object and it'd tell us what the "biggest" row is.

Now - I suppose, a bigger problem might be - what happens if we have one row that's big and many tiny ones? I don't have a solution for this.

nik9000 replied:
I suppose a way to do this is to copy the built object into a second one that's perfectly sized if we're much too big.

dnhatn replied:
Can we maintain a `lastInsertedRow` and use it as a reference for the new row? We can pass this row as a parameter to the constructor of the Row. This way, we can initialize the keys and values with the sizes of the previous keys and values. I also think we should limit the initial sizes to avoid a big row with many tiny rows cases.

alex-spies replied:
Yeah, using the last row as indicator may be super-duper off, e.g. if we alternate between short and long rows - then we're always off. I'm not sure if using a `lastInsertedRow` would solve that. (The inserted rows sizes might not be indicative of the rows we still encounter and thus have to allocate for, but I may be wrong.)

I like the idea of copying out into a perfectly sized row. This way we can keep a single row as buffer, let it grow to whatever max size we need, and only discard it after the operator has run - this will minimize allocations even more.

nik9000 replied:
If we copied to a perfectly sized row I'd lower the max target - like half it or something. Or just set it back to the perfectly sized row. Then the next row is still perfectly sized. It'd work great for runs of big rows and runs of small rows. Much less good for big row/small row/big row/small row. But that feels less likely than runs.

alex-spies commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/topn/TopNOperator.java:
```diff
@@ -273,6 +276,12 @@ public String describe() {
+     * Esp. in case of many fields, a row's values can become quite large; growing from an empty row every time is expensive due
+     * to lots of allocations. Instead, keep track of the last row's actual size and use that to pre-allocate memory for the next row.
+     */
+    private int lastSeenRowValuesSize = 32;
```
Maybe this should start at 0 - in case that the sort keys are all we have, we might be unnecessarily allocating data for the values.

nik9000 replied:
I'm good starting it at 0, yeah. I could absolutely see a world where we're only sorting, like, a stream of `int`s or `keyword`s.

alex-spies committed:
Update docs/changelog/104796.yaml

elasticsearchmachine commented:
Hi @alex-spies, I've created a changelog YAML for you.

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

alex-spies commented:
To be on the safe side, I ran the nyc_taxis benchmark (with `ingest_percentage: 5`) and didn't see performance regressions, so I'm taking this out of draft mode.

nik9000 commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/topn/TopNOperator.java:
```diff
@@ -71,12 +71,15 @@ static final class Row implements Accountable, Releasable {
             boolean success = false;
             try {
                 keys = new BreakingBytesRefBuilder(breaker, "topn");
                 values = new BreakingBytesRefBuilder(breaker, "topn");
```
I'd make the constructor take a capacity here. There's a constructor for `BytesRef` that takes a capacity too. Honestly there are only 7 calls to this so I'd just add it to the existing ctor.

dnhatn replied:
++

nik9000 reviewed (commented):
It'd be nice if we could count the number of resizes somehow so we had something that'd come out of the test to prove that this change "sticks". It feels a little like a waste of time. But I tend to do it just in case. Tests really are for "me in size months" - when I've forgotten *why* I made some change.

alex-spies committed:
Preallocation size in BreakingBytesRefBuilder ctor

alex-spies committed:
Test shrinkToFit

alex-spies committed:
Pre-allocate row's keys as well

alex-spies committed:
Implement BreakingBytesRefBuilder::shrunkCopy

alex-spies committed:
Revert "Implement BreakingBytesRefBuilder::shrunkCopy"

This reverts commit 91c06d4baabef24c191891ecec775d9f78d9c562.

alex-spies committed:
Do not shrink, use max row size to pre-allocate

alex-spies committed:
Assert num resizes in a test

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into improve-perf-topn-many-large-fields

alex-spies commented:
Thanks for the reviews, Nik and Nhat!

I included a test that asserts the number of resizes happening during execution of the TopN operator. For now, this is estimating the pre-allocation based on the max row size it has seen. In the test I wrote, that brings down the number of resize events to 61 from 450-ish.

alex-spies committed:
Do not request memory twice from breaker in ctor

alex-spies committed:
Fix wrong preallocation...

alex-spies committed:
Use constant empty bytes ref in case of 0 capacity

alex-spies committed:
Cleanup

dnhatn commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/BreakingBytesRefBuilder.java:
```diff
@@ -25,12 +25,27 @@ public class BreakingBytesRefBuilder implements Accountable, Releasable {
+    /**
+     * Counts how often we needed to resize while growing. Only works if assertions are enabled.
+     */
+    private int resizes = 0;
```
I prefer not to have these 4 bytes to save memory. It's fine not to test the number of resizes. Or we can track these numbers with a custom circuit breaker in tests.

alex-spies replied:
Custom circuit breaker is a great idea, I'll go with that!

dnhatn commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/topn/TopNOperator.java:
```diff
@@ -273,6 +273,14 @@ public String describe() {
+    /**
+     * Counts how often we needed to resize rows. Only works if assertions are enabled.
+     */
+    private int rowResizes;
```
This is not about the memory usage, but I think this class is simpler without it. We can test the number of resizes with a custom circuit breaker in tests.

dnhatn commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/topn/TopNOperator.java:
```diff
@@ -349,12 +357,22 @@ public void addInput(Page page) {
+                // When rows are very long, appending the values one by one can lead to lots of allocations.
+                // To avoid this, pre-allocate at least as much size as in the last seen row.
+                // Let the pre-allocation size decay in case we only have 1 huge row and smaller rows otherwise.
+                spareKeysPreAllocSize = Math.max(spare.keys.length(), spareKeysPreAllocSize / 2);
```
Nice!

nik9000 replied:
I think this is a better compromise than I was thinking. I like the idea of a buffer, but we don't have time for that at the moment.

dnhatn reviewed (approved):
I left a comment regarding tracking the number of resizes. I prefer not to add it in production code; otherwise, LGTM. Thanks, Alex!

alex-spies committed:
Use custom circuit breaker to count allocations

alex-spies commented:
Alright, added `CountingCircuitBreaker` in aa04c9ee2bdc085f93f1d4a0b17904cdaedfcc4c, which can track overall allocations in tests. Using it requires much less wiring and works just as well - good idea Nhat!

(In the test I added, allocations go down from 484 to 94 with the optimization from this PR.)
